userspace/libsinsp/metrics_collector.cpp

```diff
@@ -506,90 +474,267 @@ void libs_metrics_collector::snapshot()
 		}
 	}
 
+	/* We need to declare here all variables since they are captured by reference by the lambdas */
+	// METRICS_V2_RESOURCE_UTILIZATION related
+	uint32_t rss{0}, vsz{0}, pss{0}, host_procs_running{0};
+	uint64_t host_memory_used{0}, host_open_fds{0};
+	double cpu_usage_perc{0.0}, host_cpu_usage_perc{0.0};
+	uint64_t container_memory_used = get_container_memory_used();
+
+	// METRICS_V2_STATE_COUNTERS related
+	uint64_t n_fds = 0;
+	uint64_t n_threads = 0;
+	std::shared_ptr<const sinsp_stats_v2> sinsp_stats_v2 = m_inspector->get_sinsp_stats_v2();
+
+	const std::function<metrics_v2()> sinsp_stats_v2_collectors[] = {
+		[SINSP_RESOURCE_UTILIZATION_CPU_PERC] = [this,&cpu_usage_perc]() {
```

Comment:
This is using C99 designated initializers - available in C++ only through a gcc extension - in a C++ context. The compiler allows us to get away with this, so I can live with it and throw away PR #1693 that was supposed to overcome this stretched usage of the language.
It's just that it feels like a lot of machinery to fill an array (`sinsp_stats_v2_collectors`), that will be used to fill a vector (`m_metrics`), that is used only in tests, IIUC; at least for the moment, there's probably more use of this lying ahead.

I don't have a solution, but it seems to me there ought to be a simpler way to reach the goal. Have basic C99 designated initializers for structs been tried?
Something like this.

```c++
  metrics_v2 sinsp_stats_v2_collectors[] = {
      [SINSP_RESOURCE_UTILIZATION_CPU_PERC] = {
          .name = "cpu_usage_perc",
          .flags = METRICS_V2_RESOURCE_UTILIZATION,
          .metric_type = METRIC_VALUE_METRIC_TYPE_NON_MONOTONIC_CURRENT,
          .value = { .u32 = cpu_usage_perc },
          .type = METRIC_VALUE_TYPE_D,
          .unit = METRIC_VALUE_UNIT_PERC
      },
      // ...
  };
```

Perhaps it looks a tad verbose, but it is pure C, hence it can stay in a separate C file where this construct is legal; it needs no `new_metric` template function, no need for lambdas used as factories, and it would represent all the information on metrics in a sort of "tabular" format, as if it were a metrics "data sheet".

Reply:
> It's just that it feels like a lot of machinery to fill an array (sinsp_stats_v2_collectors), that will be used to fill a vector (m_metrics), that is used only in tests, IIUC; at least for the moment, there's probably more use of this lying ahead.

It's all used in Falco, see for example this PR https://github.com/falcosecurity/falco/pull/3129 where I bump libs.

I probably can't add a lot of value here, other than saying that I liked having what you @federico-sysdig described as 

> information on metrics in a sort of "tabular" format, as if it were a metrics "data sheet".

Reply:
> Perhaps it looks a tad verbose, but it is pure C, hence it can stay in a separate C file where this construct is legal; it needs no new_metric template function, no need for lambdas used as factories, and it would represent all the information on metrics in a sort of "tabular" format, as if it were a metrics "data sheet".

While i agree this would be a simple yet descriptive table, i don't get how would one link in the proper value: in your example your are copying `cpu_usage_perc` by value; does it mean that you expect the table to be defined **after** all metrics are collected? 
Or would we use pointers for the value?

Reply:
Oh, the lambda is to get a value that dynamically changed. Now I see it. What was wrong with the original implementation, then? The fact that there was no explicit link between the index that identifies the metric and its initialization in the vector?

Reply:
> Oh, the lambda is to get a value that dynamically changed

Yep :)
> What was wrong with the original implementation, then?

The fact that we had no guarantee that a metric whose name was declared in the initial array was then actually used and pushed in code below; now we have that guarantee. 
While it might seem "minor", considering the number of metrics we already have (plus more in the future perhaps?) it gets super error prone.

Reply:
Yes, I suspected that. I'd suggest to view this code as something that could be improved up, if the right idea comes up. As I wrote, it feels a bit involved, but as I don't have a solution right away I can only express my gut feeling about it.

Reply:
>  I'd suggest to view this code as something that could be improved up, if the right idea comes up

I agree; i don't really like using fancy C++ stuff for this "small" thing; at the same time, i am pretty sure that sooner or later we are going to touch this code to add new metrics and we'll get lost through all the `emplace_back(new_metric(`. @sgaist perhaps has more ideas?

Reply:
Not that I want to add complexity to the build system but I am wondering whether we should consider code generation to create this table ?
Have one file that encodes all the parameters that defines a metrics and then something parsing it and generating the adequate C code to build it.

Reply:
I think the problem would still be how to "capture" the referenced value without doing weird stuff.
Also, yep not a big fan of adding build time complexity!

Reply:
cc @jasondellaluce for more ideas :)

Reply:
Any consensus re what's an improvement vs what would be too much? Thank you!

Reply:
Kind bump re consensus :) thanks.